Wikipedia/View Controllers/WebView/NativeSectionTitleOverlays/WMFSectionTitlesViewController.m

```diff
@@ -0,0 +1,231 @@
+            if (m.anchor && m.topConstraint) {
+                NSNumber* topOffset = headingsTopOffsets[i];
+                CGFloat topFloat    = topOffset.floatValue + self.webView.scrollView.contentOffset.y;
+                m.topConstraint.constant = topFloat;
+                m.yOffset                = topFloat;
+            }
+        }
+    }
+}
+
+- (void)updateTopOverlayForScrollOffsetY:(CGFloat)offsetY {
+    [self updateIndexOfNativeTitleLabelNearestTopForScrollOffsetY:offsetY];
+    [self nudgeTopStaticTitleLabelIfNecessaryForScrollOffsetY:offsetY];
+}
+
+- (void)updateIndexOfNativeTitleLabelNearestTopForScrollOffsetY:(CGFloat)offsetY {
+    self.topStaticNativeTitleLabel.alpha = (offsetY <= 0) ? 0 : 1;
+
+    CGFloat lastOffset          = 0;
+    NSUInteger newTopLabelIndex = -1;
+
+    for (NSUInteger thisIndex = 0; thisIndex < self.models.count; thisIndex++) {
+        WMFTitleOverlayModel* m = self.models[thisIndex];
+        CGFloat thisOffset      = m.yOffset;
+        if (offsetY > lastOffset && offsetY <= thisOffset) {
+            newTopLabelIndex = thisIndex - 1;
+            break;
+        } else if ((thisIndex == (self.models.count - 1)) && (offsetY > thisOffset)) {
+            newTopLabelIndex = thisIndex;
+            break;
+        }
+        lastOffset = thisOffset;
+    }
+
+    if (newTopLabelIndex != -1) {
+        if (newTopLabelIndex != self.indexOfNativeTitleLabelNearestTop) {
+            self.indexOfNativeTitleLabelNearestTop = newTopLabelIndex;
+            [self updateTopStaticTitleLabelText];
+        }
+    }
+}
+
+- (void)nudgeTopStaticTitleLabelIfNecessaryForScrollOffsetY:(CGFloat)offsetY {
+    NSUInteger pusherIndex  = self.indexOfNativeTitleLabelNearestTop + 1;
+    CGFloat distanceToPushY = 0;
+    if (pusherIndex < (self.models.count)) {
+        WMFTitleOverlayModel* pusherTitleLabel = self.models[pusherIndex];
+        if (pusherTitleLabel.sectionId > 0) {
+            CGFloat topmostHeaderOffsetY  = pusherTitleLabel.yOffset;
+            CGRect staticLabelPseudoRect  = CGRectMake(0, 0, 1, self.topStaticNativeTitleLabel.frame.size.height);
+            CGRect topmostLabelPseudoRect = CGRectMake(0, topmostHeaderOffsetY - offsetY, 1, 1);
+            if (CGRectIntersectsRect(staticLabelPseudoRect, topmostLabelPseudoRect)) {
+                distanceToPushY = staticLabelPseudoRect.size.height - topmostLabelPseudoRect.origin.y;
+            }
+        }
+    }
+    self.topStaticNativeTitleLabelTopConstraint.constant = -distanceToPushY;
+}
+
+- (void)updateTopStaticTitleLabelText {
+    self.topStaticNativeTitleLabel.alpha = 1.0;
+    WMFTitleOverlayModel* m = self.models[self.indexOfNativeTitleLabelNearestTop];
+    self.topStaticNativeTitleLabel.text      = m.title;
+    self.topStaticNativeTitleLabel.sectionId = m.sectionId;
+}
+
+#pragma mark Section title and title location determination
+
+- (id)getJSONFromWebViewUsingFunction:(NSString*)jsFunctionString {
```

Comment:
Actually this and all the methods below seem like more of a toolkit that should be in a more centralized js place. Do we have a place where we keep our "js strings" and this type of logic now?